velox/exec/Window.cpp

```diff
@@ -379,25 +442,73 @@ void Window::callApplyForPartitionRows(
         }
         break;
       }
-      case core::WindowNode::BoundType::kPreceding:
-      case core::WindowNode::BoundType::kFollowing:
-        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kPreceding: {
+        if (frameArg->constantValue != kConstantChannel) {
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow - frameArg->constantValue - firstPartitionRow);
+        } else {
+          readFrameValues(frameArg.value());
+          for (auto i = 0; i < numRows; i++) {
+            auto offset = frameArg->columnarOffset->valueAt(i);
+            rawFrameBounds[i] = (startRow + i) - offset - firstPartitionRow;
+          }
+        }
+        break;
+      }
+      case core::WindowNode::BoundType::kFollowing: {
+        if (frameArg->constantValue != kConstantChannel) {
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow + frameArg->constantValue - firstPartitionRow);
+        } else {
+          readFrameValues(frameArg.value());
+          for (auto i = 0; i < numRows; i++) {
+            auto offset = frameArg->columnarOffset->valueAt(i);
+            rawFrameBounds[i] = (startRow + i) + offset - firstPartitionRow;
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  auto fixOutOfBoundFrames = [&](vector_size_t* frameStart,
+                                 vector_size_t* frameEnd) -> void {
+    for (auto i = frameStart, j = frameEnd; i < frameStart + numRows;
+         i++, j++) {
+      if (*i <= *j) {
+        *i = std::max(*i, 0);
+        *j = std::min(*j, lastPartitionRow - firstPartitionRow);
```

Comment:
lastPartitionRow - firstPartitionRow can be computed once and passed to this function. It is the last row of the partition.